### PR TITLE
Fix/Finish IADS

### LINKED ISSUE #2253: Not all map features updating properly
### Affected versions

Development build

### Development build

_No response_

### Description

I found that many functions on the map are not updating properly. Things like the `ThreatZones` and `NavMeshes` in the left menu only update once, after that it seems to be getting the same data over and over again. Options on right side like "Ships", "Factories", "Air defences", etc. seem to be FUBAR as well. Also supply routes were not being drawn in the correct color after a capture event. So yeah, I think I found a new issues to solve 🤣 

An overview:

- [x] ThreatZones not updating after first update
- [x] Navmeshes not updating after first update
- [x] SupplyRoutes not updating at all
- [x] Air defences: FUBAR
- [x] Factories: FUBAR
- [x] Ships : FUBAR
- [x] Other ground objects: FUBAR
- [x] Active combat: doesn't remove lines when combat ends (dead units should be erased after some period?)
- [x] Allied/Enemy IADS network: doesn't seem to do anything, using some debug messages I can see connections that are linked with themselves

### Save game and other files

Any campaign should do since it's mostly front-end stuff...

### LINKED ISSUE #2263: Push-based events overhaul
### Affected versions

Development build

### Development build

_No response_

### Description

Opening this issue to keep better track of the overhaul, labeling it (performance) bug/UX because fixing this will improve the UX since the map will update faster, more towards what we're used to from v5.2. Consider this as the first phase, effectively being a clean-up of `eventstream.tsx` to eliminate get-requests that are sent from the front-end as much as possible:

- [x] Navmesh updates (#2254)
- [x] Unculled zones (#2264)
- [x] Threat zones (#2254)
- [x] Updated flights (#2265)
- [x] Updated front lines (#2266)
- [x] Updated Tgos (#2269)
- [x] Updated IADS (#2262)
- [x] Updated control points (#2267)
- [x] Updated supply routes (#2255)
- [x] Combat updates: redundancies present (#2268)

Did I miss anything?

### Save game and other files

N/A

### PR DESCRIPTION
**Pull request by Raffson**
It's a big one, I'll write down comments in most parts of the code to explain the reasoning behind it all.

Partial fix #2253
Partial fix #2263

### CODE REVIEW

File: client/src/api/eventstream.tsx
Comment (Raffson) [inline client/src/api/eventstream.tsx:172]: All changes in this file should be obvious-ish, we're trying to decouple it from TGOs and listen for update or delete events. Initialization is done somewhere else and I notice no issues so I didn't touch that part.

File: game/theater/controlpoint.py
Comment (Raffson) [inline game/theater/controlpoint.py:846]: Readability improvement, verify it does the same as the code before...

File: game/theater/controlpoint.py
Comment (Raffson) [inline game/theater/controlpoint.py:851]: This is a new part, needed to remove connections to a comms tower or power station. If I missed anything, it's probably doing the same update (as in deleting connections) whenever a comms tower or power station are destroyed, though I'm not sure if that's being taken care of already.

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py]: Push update events whenever a connection is added for a node...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:132]: Simply dropping the entire try-except block, the only exception that could be triggered is in `dcs_name_for_group` on line 54, which imo is an exception we want to know about because if I understood the code, it's a case that shouldn't be possible.

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py]: Delete the current connections for the node...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:169]: Replacing the exception, line 169 is also new, it was missing functionality for connecting the new node to the network.

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:181]: No need for this, we can act accordingly when we check for None. If we would need to raise exceptions, we do it elsewhere, specifically where it wouldn't be considered normal behavior...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:241]: Ought to guarantee the same behavior...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:275]: Another rewrite for readability that should do exactly the same as before...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:265]: Refactor this into `_connect_to_network` since the logic is sound and I need to reuse this...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:287]: For readability, explains the addition to `iadsrole.py`

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:307]: Must also check if it's friendly, else no connection should be made...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py]: This is another situation of which I believe we should be notified, because this shouldn't be possible to occur...

Comment (Raffson) [reply game/theater/iadsnetwork/iadsnetwork.py]: Well, mypy complains this code is unreachable, I seriously hope it's right because the only way to fix it is to delete it -_-

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [reply game/theater/iadsnetwork/iadsnetwork.py]: About to delete this cause we're passing things to an events object now...

File: game/theater/theatergroundobject.py
Comment (Raffson) [inline game/theater/theatergroundobject.py:628]: This restricts flights types to Strike and DEAD only! Just consider it a building like any other because while it's a part of the entire air defense system, a factory or a comms tower on their own are not an air defense...

File: game/theater/controlpoint.py
Comment (Raffson) [inline game/theater/controlpoint.py:844]: Should still be doing the same, just a readability improvement...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:145]: Case for power suppliers and comm towers

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [inline game/theater/iadsnetwork/iadsnetwork.py:175]: Instead of raising an exception, we consider a tgo not to be participating if None is returned

File: game/theater/theatergroundobject.py
Comment (DanAlbert) [reply game/theater/theatergroundobject.py:628]: I'm guessing the intent was to exclude BAI rather than to exclude all the escort tasks and similar. This makes BAI valid again which is wrong. I'm not sure what the right fix is off the top of my head, but we shouldn't merge this. This is what I mean when I keep saying "We can't merge half a PR". All the front-end changes appear fine but other parts need to be changed (I'm not sure if the front-end changes are coupled to the back-end parts that have review comments though, so maybe I wouldn't have been able to merge anyway).

File: game/theater/controlpoint.py
Comment (DanAlbert) [reply game/theater/controlpoint.py:844]: This is a behavior change because `ShipGroundObject` now counts. Maybe that was intentional, but your comment says otherwise (this is what I mean when I say "multi-task PRs are hard to review", I'm not sure what the _intent_ of either change is)

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (DanAlbert) [reply game/theater/iadsnetwork/iadsnetwork.py:145]: If this is non-obvious it should be a _code_ comment. No one will read the PR to figure out what this function is doing after it's merged.

I'm not actually sure what the PR comment means either. Why is it only relevant for those types of nodes and not others?

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (DanAlbert) [inline game/theater/iadsnetwork/iadsnetwork.py:138]: This is impossible. Fix the signature.

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (DanAlbert) [inline game/theater/iadsnetwork/iadsnetwork.py:144]: This seems upside-down... is there not a way to iterate connections directly on the tgo?

**Review by DanAlbert — CHANGES_REQUESTED** [review]
You have a few things going on here:

1. Fix missing IADS update events that the front-end should be receiving
2. Change the mission types that are allowed on IADS nodes
3. Include more TGO types (Ships, at least) in the definition of IADS
4. Fix... something about updating the IADS network in the backend after some sort of event.

Part 1 is mergeable but I can't do that because parts 2 and 3 have review comments, and I'm not able to review part 4 at all because I don't know how this system works. Please split up your PRs by task.

File: game/theater/controlpoint.py
Comment (Raffson) [reply game/theater/controlpoint.py:844]: You're right, but look at iadsrole.py, clearly shows that naval units act as EWR in the network so indeed it is not doing the same in the sense that it includes a fix for naval units...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [reply game/theater/iadsnetwork/iadsnetwork.py:144]: Comm towers and power stations have no connections, they're not nodes. Nodes connect to them instead...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [reply game/theater/iadsnetwork/iadsnetwork.py:145]: Comm towers and power stations have no connections, they're not nodes. Nodes connect to them instead... you should ask @RndName why he designed it this way, but if I had to guess I think it's to prevent storing and drawing connections twice...

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [reply game/theater/iadsnetwork/iadsnetwork.py:138]: You mean drop the guard?

File: game/theater/iadsnetwork/iadsnetwork.py
Comment (Raffson) [reply game/theater/iadsnetwork/iadsnetwork.py]: Replaced by passing events along

File: game/theater/controlpoint.py
Comment (Raffson) [reply game/theater/controlpoint.py:844]: Ok I realize I need to rework this, because the iads update has to happen for all iads units connected to the CP, but comm towers and power stations shouldn't rotate as compared to a sam/ewr site...

Comment (Raffson) [reply game/theater/controlpoint.py:844]: Resolving this as I have corrected this part, will become available in the next push...

**Pull request closed without merging**
